service/matching/matcher_test.go

```diff
@@ -210,55 +210,59 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 	t.Equal(mustParent(t.queue.partition.(*tqid.NormalPartition), 20).RpcName(), req.GetTaskQueue().GetName())
 }
 
+//nolint:errcheck
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
-	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	intruptC := make(chan struct{})
 
-	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	// task waits for a local poller
+	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
 
-	intruptC := make(chan struct{})
-	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
+	var wg sync.WaitGroup
+	go func() {
+		t.rootMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:revive
+	}()
 
-	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
-	t.Nil(err)
-	t.False(happened) // sync match did not happen, but we called the forwarder client
+	// Wait for the task to be added to the map
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			if len(t.rootMatcher.backlogTasksCreateTime) > 0 {
+				return // Exit if a task has been added
+			}
+		}
+	}()
+	wg.Wait()
 
 	// should not allow sync match when there is an old task in backlog
-	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
-	go t.childMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
+	backlogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
+	backlogTask.forwardInfo = &taskqueuespb.TaskForwardInfo{
+		SourcePartition:    "",
+		TaskSource:         0,
+		RedirectInfo:       nil,
+		DispatchBuildId:    "",
+		DispatchVersionSet: "",
+	}
+	newCtx, newCtxCancel := context.WithTimeout(context.Background(), 1*time.Second)
+	// When the root partition has no pollers and is offered a task from the backlog, the task
```

Comment:
```suggestion
	// When the root partition has no pollers and is offered a task for sync match, the task
```

The comment makes sense otherwise 👍